### PR TITLE
Chain without recursion

### PR DESCRIPTION
**Pull request by anders-rasmussen**
The current `Parse.Chain()` operator is implemented using recursion. When applied to expressions with moderate size chain of operands and operators, this can cause a stack-overflow.

This pull request changes the implementation of `Parse.Chain()` to an iterative approach, avoiding stack-overflow exceptions.

Attached unit test show how to provoke the stack overflow exception.

### CODE REVIEW

**Comment by anders-rasmussen** [comment]
Rebased against dev branch

File: src/Superpower/Parse.cs [hunk 1]
```diff
@@ -41,22 +41,29 @@ public static class Parse
...
+                }
+
+                var result = parseResult.Value;
+
+                var operatorResult = @operator(parseResult.Remainder);
+                while (operatorResult.HasValue) {
+                    parseResult = operand(operatorResult.Remainder);
+
+                    if (!parseResult.HasValue) {
+                        return Result.CastEmpty<T, T>(parseResult);
```
Comment (nblumhardt) [inline src/Superpower/Parse.cs]: This is effectively a cast from `T` to `T` - I think we can just return `parseResult` here (and in the corresponding `TokenListParser` below).

**Comment by nblumhardt** [comment]
If we change the grammar of `IntCalc` to use multi-token operators, e.g. `++` for add (two tokens), we get a change in error reporting:

Before - we've partially accepted the `+` and report the missing `+`:

```
>IntCalc.exe 1 + 3
Syntax error (line 1, column 5): unexpected number `3`, expected `+`.
```

After, we lose the partial parse and can't provide as specific information:

```
>IntCalc.exe 1 + 3
Syntax error (line 1, column 3): unexpected operator `+`.
```

I think the problem here is just that when we try parsing the operator, if it fails but `operatorResult.IsPartial` is true, we need to return the failure directly.

Here's a shot at a test case:

```csharp
var nPlusPlusN = Parse.Chain(
    Character.EqualTo('+').IgnoreThen(Character.EqualTo('+')),
    Numerics.IntegerInt32,
    (c, i, j) => i + j);

AssertParser.FailsAt(nPlusPlusN, "1+1", 3);
```

This is different from Sprache's chain, which does backtrack, in case you've encountered it. In Sprache, there are non-backtracking versions of many parsers, `XChainOperator()` behaves like the one we're discussing here, IIRC. Superpower uses a different approach, requiring opt-in to backtracking on individual operators with constructs via `Try()`,  like `Parse.Chain(x.Try()...`.

**Comment by anders-rasmussen** [comment]
True, the changes did change how error messages was handled when the operator parser failed - this was not intentional. 

I've added a check for the operator parser result being partial, to return an error in this scenario. Thanks for the test case to detect this.

I'm not familiar with Sprache - but if I understand the Superpower backtracking correctly, it only happens when one explicitly call the `Try()` (which looks at the `Backtrack` property) . Looking at how the existing `Chain` is implemented, it assume it is safe to ignore backtracking for the operand parser part.

**Comment by nblumhardt** [comment]
👍 

I think backtracking in the current `Chain()` is likely an oversight, it'd be great to squeeze it out, if we can :-)

**Comment by nblumhardt** [comment]
Apologies for the slow turnaround, looks good! 👍

**Pull request merged**